xarray/tests/test_coding_times.py

```diff
@@ -1901,3 +1907,17 @@ def test_lazy_decode_timedelta_error() -> None:
     )
     with pytest.raises(OutOfBoundsTimedelta, match="overflow"):
         decoded.load()
+
+
+@pytest.mark.parametrize("decode_timedelta", [True, False])
+@pytest.mark.parametrize("mask_and_scale", [True, False])
+def test_decode_timedelta_mask_and_scale(
+    decode_timedelta: bool, mask_and_scale: bool
+) -> None:
+    attrs = {"units": "days", "_FillValue": np.int16(-1), "add_offset": 100.0}
```

Comment:
Since the offset is so small relative to `np.iinfo(np.int64).min`, I think this currently may miss a missing-value-related bug.  This test should be more challenging:
```suggestion
    attrs = {"units": "nanoseconds", "_FillValue": np.int16(-1), "add_offset": 100000.0}
```

Reply:
Great, this really breaks the current approach. I'll have to rework the encoding part, too.

Reply:
No worries!  If any of [this kind of casting](https://github.com/pydata/xarray/blob/2475d4965c8b3f69f70b0acc958fa98ce4a45351/xarray/coding/times.py#L1237-L1239) is problematic, feel free to take a look at https://github.com/pydata/xarray/pull/9498.  I synced it with `main`, and should be ready for review.

Reply:
Thanks for the heads-up. I've tried to be consistent now in the encoding and decoding part. Would be great to get another look here :-)